rasterio/_io.pyx

```diff
@@ -973,6 +976,84 @@ cdef class DatasetReaderBase(DatasetBase):
 
         return out
 
+    def read_rat(self, bidx):
+        """Read a raster attribute table (rat) for a band from the dataset.
+
+        Parameters
+        ----------
+        bidx : int
+            Index of the band whose raster attribute table will be returned.
+            Band index starts at 1.
+
+        Returns
+        -------
+        dict
+            Mapping of column name to a numpy array of column values
+        dict
+            Mapping of column name to field usage, and a table type enum.
+        GDALRATTableType.<enum>
+            Raster attribute table type (thematic or athematic)
+
+        Raises
+        ------
+        ValueError
+            If no raster attribute table is found for the specified band.
+        ValueError
+            If a column data type is not understood.
+        IndexError
+            If no band exists for the provided index.
+        """
+
+        cdef GDALRasterBandH band = NULL
+        cdef GDALRasterAttributeTableH rat = NULL
+
+        band = self.band(bidx)
+        rat = GDALGetDefaultRAT(band)
+
+        if rat == NULL:
+            raise ValueError("NULL raster attribute table")
+        
+        row_count = GDALRATGetRowCount(rat)
+        col_count = GDALRATGetColumnCount(rat)
+
+        retval = {}
+        retuse = {}
+
+        for c in range(col_count):
+            col_name = GDALRATGetNameOfCol(rat, c).decode('utf-8')
+            col_type = GDALRATGetTypeOfCol(rat, c)
+            col_use  = GDALRATGetUsageOfCol(rat, c)
+
+            if col_type == GFT_Integer:
+                values = np.array(
+                    [GDALRATGetValueAsInt(rat, r, c) for r in range(row_count)],
+                    dtype=np.int32
+                )
+            elif col_type == GFT_Real:
+                values = np.array(
+                    [GDALRATGetValueAsDouble(rat, r, c) for r in range(row_count)],
+                    dtype=np.float64
+                )
+            elif col_type == GFT_String:
+                values = [GDALRATGetValueAsString(rat, r, c) for r in range(row_count)]
+                maxlen = max(len(v) for v in values)
+
+                values = np.array(
+                    values,
+                    dtype=f'S{maxlen}'
```

Comment:
Computing maxlen is unnecessary here. We can just set the dtype here as `dtype='str'` instead of casting.